File: service/src/test/java/no/nav/melosys/service/dokument/DokumentproduksjonsInfoMapperTest.java
package no.nav.melosys.service.dokument;

import java.util.HashSet;
import java.util.Set;

import no.finn.unleash.FakeUnleash;
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter;
import no.nav.melosys.exception.FunksjonellException;
import org.junit.jupiter.api.Test;

import static no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.*;
import static no.nav.melosys.service.dokument.DokumentproduksjonsInfoMapper.DOKUMENTMALER_PRODSATT;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

class DokumentproduksjonsInfoMapperTest {
    private final FakeUnleash fakeUnleash = new FakeUnleash();

    private final DokumentproduksjonsInfoMapper dokumentproduksjonsInfoMapper = new DokumentproduksjonsInfoMapper(fakeUnleash);

    @Test
    void feilerNårMalIkkeFinnesIDokgen() {
        assertThatThrownBy(() -> dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(ATTEST_A1))
            .isInstanceOf(FunksjonellException.class)
            .hasMessage("ProduserbartDokument ATTEST_A1 er ikke støttet");
    }

    @Test
    void skalUtledeTilgjengeligeMaler() {
        Set<Produserbaredokumenter> produserbaredokumenter = dokumentproduksjonsInfoMapper.utledTilgjengeligeMaler();

        Set<Produserbaredokumenter> forventetProduserbaredokumenter = new HashSet<>(DOKUMENTMALER_PRODSATT);
        assertThat(produserbaredokumenter).hasSameElementsAs(forventetProduserbaredokumenter);
    }

    @Test
    void skalHenteDokumentInfo() {
        DokumentproduksjonsInfo dokumentproduksjonsInfo = dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(MANGELBREV_BRUKER);

        assertThat(dokumentproduksjonsInfo.dokgenMalnavn()).isEqualTo("mangelbrev_bruker");
        assertThat(dokumentproduksjonsInfo.dokumentKategoriKode()).isEqualTo("IB");
        assertThat(dokumentproduksjonsInfo.journalføringsTittel()).isEqualTo("Melding om manglende opplysninger");
    }

    @Test
    void skalHenteMalnavn() {
        String malnavn = dokumentproduksjonsInfoMapper.hentMalnavn(MELDING_FORVENTET_SAKSBEHANDLINGSTID);
        assertThat(malnavn).isEqualTo("saksbehandlingstid_soknad");
    }
}

File: service/src/test/java/no/nav/melosys/service/dokument/brev/mapper/DokumentproduksjonsInfoMapperTest.kt
package no.nav.melosys.service.dokument.brev.mapper

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.*
import no.nav.melosys.exception.FunksjonellException
import org.junit.jupiter.api.Test

class DokumentproduksjonsInfoMapperTest {

    private val dokumentproduksjonsInfoMapper = DokumentproduksjonsInfoMapper()

    @Test
    fun `feiler når mal ikke finnes i dokgen`() {
        val exception = shouldThrow<FunksjonellException> {
            dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(ATTEST_A1)
        }
        exception.message shouldContain "ProduserbartDokument ATTEST_A1 er ikke støttet"
    }

    @Test
    fun `skal hente dokumentinfo`() {
        val dokumentproduksjonsInfo = dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(MANGELBREV_BRUKER)


        dokumentproduksjonsInfo.run {
            dokgenMalnavn shouldBe "mangelbrev_bruker"
            dokumentKategoriKode shouldBe "IB"
            journalføringsTittel shouldBe "Melding om manglende opplysninger"
        }
    }

    @Test
    fun `skal hente malnavn`() {
        val malnavn = dokumentproduksjonsInfoMapper.hentMalnavn(MELDING_FORVENTET_SAKSBEHANDLINGSTID_SOKNAD)


        malnavn shouldBe "saksbehandlingstid_soknad"
    }
}
